src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -189,6 +198,137 @@ internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projec
             return builder.ToImmutable();
         }
 
+        /// <summary>
+        /// Computes a new <see cref="_lazyReverseReferencesMap"/> for the removal of a project.
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseReferencesMapForRemovedProject(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingForwardReferencesMap,
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReverseReferencesMap,
+            ProjectId projectId)
+        {
+            var builder = existingReverseReferencesMap.ToBuilder();
+
+            // Iterate over each project referenced by 'projectId', which is now being removed. Update the reverse
+            // references map for the project to no longer include 'projectId' in the list.
+            if (existingForwardReferencesMap.TryGetValue(projectId, out var referencedProjectIds))
+            {
+                foreach (var referencedProjectId in referencedProjectIds)
+                {
+                    if (builder.TryGetValue(referencedProjectId, out var reverseDependencies))
+                    {
+                        builder[referencedProjectId] = reverseDependencies.Remove(projectId);
+                    }
+                }
+            }
+
+            // Finally, remove 'projectId' itself.
+            builder.Remove(projectId);
+            return builder.ToImmutable();
+        }
+
+        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseReferencesMapForRemovedProjectReference(
```

Comment:
I'm not sure how this file was ordered before but I kinda wish we grouped by "here's all the methods for project removal", "here's all the methods for project reference removal".

Reply:
(Heck, I might even like partial classes with each operation and the top-level public method as well.)

Reply:
i like partial types for just this purpose.  :)

Reply:
Split locally; will mark resolved on push

Reply:
➡️ Now pushed.